File: Services/pkg/permission/service.go
package permission

import (
	"context"

	"github.com/ChangerzaryX1602/SkillSync/pkg/domain"
	"github.com/ChangerzaryX1602/SkillSync/pkg/models"

	helpers "github.com/zercle/gofiber-helpers"
)

type permissionService struct {
	repository domain.PermissionRepository
}

func NewPermissionService(repository domain.PermissionRepository) domain.PermissionService {
	return &permissionService{repository: repository}
}

func (s *permissionService) CreatePermission(ctx context.Context, permission models.Permission) []helpers.ResponseError {
	err := s.repository.CreatePermission(ctx, permission)
	if err != nil {
		return []helpers.ResponseError{*err}
	}
	return nil
}

func (s *permissionService) GetPermission(ctx context.Context, id uint) (*models.Permission, []helpers.ResponseError) {
	permission, err := s.repository.GetPermission(ctx, id)
	if err != nil {
		return nil, []helpers.ResponseError{*err}
	}
	return permission, nil
}

func (s *permissionService) GetPermissions(ctx context.Context, pagination models.Pagination, search models.Search) ([]models.Permission, *models.Pagination, *models.Search, []helpers.ResponseError) {
	permissions, paginated, searched, err := s.repository.GetPermissions(ctx, pagination, search)
	if err != nil {
		return nil, nil, nil, []helpers.ResponseError{*err}
	}
	return permissions, paginated, searched, nil
}

func (s *permissionService) UpdatePermission(ctx context.Context, id uint, permission models.Permission) []helpers.ResponseError {
	err := s.repository.UpdatePermission(ctx, id, permission)
	if err != nil {
		return []helpers.ResponseError{*err}
	}
	return nil
}

func (s *permissionService) DeletePermission(ctx context.Context, id uint) []helpers.ResponseError {
	err := s.repository.DeletePermission(ctx, id)
	if err != nil {
		return []helpers.ResponseError{*err}
	}
	return nil
}

File: hono-clean-architecture/pkg/permission/service.ts
import { Permission, Pagination, Search } from "../models";
import { PermissionRepository, PermissionService as IPermissionService } from "../domain";
import { ResponseError } from "../../internal/infrastructure/custom_error";

export class PermissionService implements IPermissionService {
  private repository: PermissionRepository;

  constructor(repository: PermissionRepository) {
    this.repository = repository;
  }

  async createPermission(permissionData: {
    group: string;
    name: string;
  }): Promise<ResponseError[]> {
    const { error } = await this.repository.createPermission(permissionData);
    if (error) {
      return [error];
    }
    return [];
  }

  async getPermission(
    id: number
  ): Promise<{ permission: Permission | null; errors: ResponseError[] }> {
    const { permission, error } = await this.repository.getPermission(id);
    if (error) {
      return { permission: null, errors: [error] };
    }
    return { permission, errors: [] };
  }

  async getPermissions(
    pagination: Pagination,
    search: Search
  ): Promise<{
    permissions: Permission[];
    pagination: Pagination;
    search: Search;
    errors: ResponseError[];
  }> {
    const result = await this.repository.getPermissions(pagination, search);
    if (result.error) {
      return { permissions: [], pagination, search, errors: [result.error] };
    }
    return {
      permissions: result.permissions,
      pagination: result.pagination,
      search: result.search,
      errors: [],
    };
  }

  async updatePermission(
    id: number,
    permissionData: Partial<Permission>
  ): Promise<ResponseError[]> {
    const error = await this.repository.updatePermission(id, permissionData);
    if (error) {
      return [error];
    }
    return [];
  }

  async deletePermission(id: number): Promise<ResponseError[]> {
    const error = await this.repository.deletePermission(id);
    if (error) {
      return [error];
    }
    return [];
  }
}

export function newPermissionService(repository: PermissionRepository): IPermissionService {
  return new PermissionService(repository);
}
